js/src/evaluation/_runner.ts

```diff
@@ -0,0 +1,932 @@
+    }
+    return this._experiment;
+  }
+
+  async _getExperimentMetadata(): Promise<KVMap> {
+    let projectMetadata = this._metadata ?? {};
+    const gitInfo = await getGitInfo();
+    if (gitInfo) {
+      projectMetadata = {
+        ...projectMetadata,
+        git: gitInfo,
+      };
+    }
+    if (this._experiment) {
+      const experimentMetadata: KVMap =
+        this._experiment.extra && "metadata" in this._experiment.extra
+          ? this._experiment.extra.metadata
+          : {};
+      projectMetadata = {
+        ...experimentMetadata,
+        ...projectMetadata,
+      };
+    }
+    return projectMetadata;
+  }
+
+  async _getProject(firstExample: Example): Promise<TracerSession> {
+    let project: TracerSession;
+    if (!this._experiment) {
+      try {
+        const projectMetadata = await this._getExperimentMetadata();
+        project = await this.client.createProject({
+          projectName: this.experimentName,
+          referenceDatasetId: firstExample.dataset_id,
+          metadata: projectMetadata,
+        });
+      } catch (e) {
+        if (String(e).includes("already exists")) {
+          throw e;
+        }
+        throw new Error(
+          `Experiment ${this._experimentName} already exists. Please use a different name.`
+        );
+      }
+    } else {
+      project = this._experiment;
+    }
+    return project;
+  }
+
+  _printExperimentStart(): void {
+    // @TODO log with experiment URL
+    console.log(`Starting evaluation of experiment: ${this.experimentName}`);
+  }
+}
+
+/**
+ * Manage the execution of experiments.
+ *
+ * Supports lazily running predictions and evaluations in parallel to facilitate
+ * result streaming and early debugging.
+ */
+class _ExperimentManager extends _ExperimentManagerMixin {
+  _data?: DataT;
+
+  _examples?: AsyncGenerator<Example>;
+
+  _runs?: AsyncGenerator<Run>;
+
+  _evaluationResults?: AsyncGenerator<EvaluationResults>;
+
+  _summaryResults?: AsyncGenerator<
+    (runsArray: Run[]) => AsyncGenerator<EvaluationResults, any, unknown>,
+    any,
+    unknown
+  >;
+
+  _examplesArray?: Example[];
+
+  _runsArray?: Run[];
+
+  constructor(args: _ExperimentManagerArgs) {
+    super({
+      experiment: args.experiment,
+      metadata: args.metadata,
+      client: args.client,
+    });
+    if (args._examplesArray && args._examplesArray.length) {
+      this._examplesArray = args._examplesArray;
+    }
+    this._data = args.data;
+
+    if (args._runsArray && args._runsArray.length) {
+      this._runsArray = args._runsArray;
+    }
+    this._runs = args.runs;
+
+    this._evaluationResults = args.evaluationResults;
+    this._summaryResults = args.summaryResults;
+  }
+
+  get examples(): AsyncGenerator<Example> {
+    if (this._examples === undefined) {
+      if (!this._data) {
+        throw new Error("Data not provided in this experiment.");
+      }
+      return _resolveData(this._data, { client: this.client });
```

Comment:
So we are fetching it but never adding to the examples Array? Why not consume and add to example array here rather than mixing and matching generators and the array?

Reply:
Could do this, only reason I didn't is because we'd have to switch all the `this.examples` calls to `await this.getExamples()` since getters can't be async. I guess in the long run it's a little cleaner though, will fix.